File: scrapers/kickass.py
"""KickAss Torrents scraper - HTML scraping with domain fallback."""

import re
from urllib.parse import quote
from scrapers.base import BaseScraper
from scrapers import client, utils


class Source(BaseScraper):
    name = 'KickAss'
    domains = [
        'kick4ss.com',
        'thekat.info',
        'kickass.cm',
        'kickass.ws',
        'kickasstorrents.to',
        'kat.am',
    ]
    base_url = 'https://kick4ss.com'

    def _find_working_domain(self):
        for domain in self.domains:
            url = f'https://{domain}'
            resp = client.get(url, timeout=5)
            if resp and resp.status_code == 200:
                return url
        return self.base_url

    def search(self, query):
        base = self._find_working_domain()
        url = f'{base}/usearch/{quote(query)}/?field=size&sorder=desc'
        soup = client.get_soup(url)
        if not soup:
            return []
        results = []
        table = soup.select_one('table.data')
        if not table:
            return []
        for row in table.select('tr')[1:]:
            cols = row.select('td')
            if len(cols) < 5:
                continue
            # Name
            name_tag = cols[0].select_one('a.cellMainLink')
            name = utils.clean_name(name_tag.get_text()) if name_tag else ''
            if not name:
                continue
            # Magnet
            magnet_tag = cols[0].select_one('a[href^="magnet:"]')
            if not magnet_tag:
                # Try to find hash from the torrent link
                link = name_tag.get('href', '')
                detail_page = client.get_soup(base + link)
                if detail_page:
                    magnet_tag = detail_page.select_one('a[href^="magnet:"]')
                if not magnet_tag:
                    continue
            magnet = magnet_tag['href']
            # Size
            size_text = cols[1].get_text(strip=True) if len(cols) > 1 else ''
            # Seeders
            seeders = utils.parse_seeders(cols[3].get_text()) if len(cols) > 3 else 0
            results.append({
                'name': name,
                'size': size_text or 'N/A',
                'size_gb': utils.parse_size(size_text),
                'seeders': seeders,
                'magnet': magnet,
                'provider': self.name,
            })
        return results

File: netlify/functions/lib/scrapers/kickass.js
/**
 * KickAss Torrents scraper - HTML scraping.
 */
const client = require('../client');
const { parseSeeders, parseSize, cleanName } = require('../utils');

const NAME = 'KickAss';
const DOMAINS = ['kick4ss.com', 'thekat.info', 'kickass.cm', 'kickasstorrents.to', 'kat.am'];

async function findWorkingDomain() {
  for (const domain of DOMAINS) {
    const resp = await client.get(`https://${domain}`, { timeout: 4000 });
    if (resp) return `https://${domain}`;
  }
  return `https://${DOMAINS[0]}`;
}

async function search(query) {
  const base = await findWorkingDomain();
  const url = `${base}/usearch/${encodeURIComponent(query)}/?field=size&sorder=desc`;
  const $ = await client.getSoup(url);
  if (!$) return [];

  const results = [];
  const table = $('table.data');
  if (!table.length) return [];

  table.find('tr').slice(1).each((_, row) => {
    const cols = $(row).find('td');
    if (cols.length < 5) return;

    const nameTag = $(cols[0]).find('a.cellMainLink').first();
    const name = cleanName(nameTag.text());
    if (!name) return;

    const magnetTag = $(cols[0]).find('a[href^="magnet:"]').first();
    if (!magnetTag.length) return;
    const magnet = magnetTag.attr('href');

    const sizeText = cols.length > 1 ? $(cols[1]).text().trim() : '';
    const seeders = cols.length > 3 ? parseSeeders($(cols[3]).text()) : 0;

    results.push({
      name,
      size: sizeText || 'N/A',
      size_gb: parseSize(sizeText),
      seeders,
      magnet,
      provider: NAME,
    });
  });
  return results;
}

module.exports = { name: NAME, base_url: `https://${DOMAINS[0]}`, search };
